src/analysis/modal_esm/predict.py

```diff
@@ -0,0 +1,222 @@
+"""Module for calculating masked position probabilities using ESM models and Modal.
+
+This module defines functions and classes to compute the probabilities at masked
+positions in sequences using ESM models. It leverages Modal for distributed GPU
+computation.
+"""
+
+from __future__ import annotations
+
+import modal
+import numpy as np
+import torch
+from modal.functions import gather
+
+from analysis.modal_esm.cache_model import app as cache_app
+from analysis.modal_esm.cache_model import download_and_cache_model
+from analysis.modal_esm.constants import (
+    CONSOLE,
+    HOURS,
+    IMAGE,
+    MINUTES,
+    MODEL_STORAGE_PATH,
+    PREDICTION_APP,
+    VOLUME,
+)
+
+with IMAGE.imports():
```

Comment:
I agree that there are issues that arise because two environments must be managed. But if I understand you correctly, I think there might be a misconception. The value of this context manager is to hide packages from local that exist only on the remote (and thus raise ImportError locally), not to specify the complete set of packages modal will use. Modal code can reference packages imported outside of the managed import block.

There are different linting issues that arise because of this though. For example, we can't lint code for packages that don't exist locally. We also can't verify that the packages referenced by any modal-wrapped functions/classes actually exist in the image.

Reply:
Ah sorry yeah that wasn't clear. What I meant was basically your second point, that there's no way to check that there are no missing imports in the functions that will run on modal. Presumably, you'd have to take the union of the actual local imports, the imports specified in this context manager, and the imports within `app.function`-wrapped functions.